File: test_tf_conversion.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
import tf2_ros
import numpy as np
from scipy.spatial.transform import Rotation as R

class TfConversionTest(Node):
    def __init__(self):
        super().__init__('tf_conversion_test')
        
        self.subscription = self.create_subscription(
            Odometry,
            '/rkbot/lio/odom',
            self.odom_callback,
            10)
        self.subscription  # prevent unused variable warning
        
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)
        
        self.get_logger().info('TF conversion test node ready')

    def odom_callback(self, msg):
        try:
            # 获取 world 到 map 的变换
            trans = self.tf_buffer.lookup_transform(
                'rkbot/map',
                'rkbot/world',
                rclpy.time.Time())
            
            self.get_logger().info('\n=== TF Conversion Test ===')
            self.get_logger().info(f'World->Map Transform:')
            self.get_logger().info(f'  Translation: [{trans.transform.translation.x:.3f}, '
                                  f'{trans.transform.translation.y:.3f}, '
                                  f'{trans.transform.translation.z:.3f}]')
            
            # 构建变换矩阵
            # T_map_world = [R  t]
            #               [0  1]
            trans_x = trans.transform.translation.x
            trans_y = trans.transform.translation.y
            trans_z = trans.transform.translation.z
            
            qx = trans.transform.rotation.x
            qy = trans.transform.rotation.y
            qz = trans.transform.rotation.z
            qw = trans.transform.rotation.w
            
            # 将四元数转换为旋转矩阵
            rot = R.from_quat([qx, qy, qz, qw])
            rot_mat = rot.as_matrix()
            
            # 4x4 变换矩阵
            T_map_world = np.eye(4)
            T_map_world[:3, :3] = rot_mat
            T_map_world[:3, 3] = [trans_x, trans_y, trans_z]
            
            # 获取 odom 在 world 坐标系下的位姿
            world_x = msg.pose.pose.position.x
            world_y = msg.pose.pose.position.y
            world_z = msg.pose.pose.position.z
            
            world_qx = msg.pose.pose.orientation.x
            world_qy = msg.pose.pose.orientation.y
            world_qz = msg.pose.pose.orientation.z
            world_qw = msg.pose.pose.orientation.w
            
            # 将 odom 位姿转换为齐次坐标
            odom_world_homogeneous = np.array([world_x, world_y, world_z, 1.0])
            
            # 应用变换：T_map_odom = T_map_world * T_world_odom
            odom_map_homogeneous = T_map_world @ odom_world_homogeneous
            
            # 计算旋转：R_map_odom = R_map_world * R_world_odom
            odom_rot_world = R.from_quat([world_qx, world_qy, world_qz, world_qw])
            odom_rot_map = rot * odom_rot_world
            odom_q_map = odom_rot_map.as_quat()
            
            self.get_logger().info(f'\nOdometry in World frame:')
            self.get_logger().info(f'  Position: [{world_x:.3f}, {world_y:.3f}, {world_z:.3f}]')
            self.get_logger().info(f'  Orientation: [{world_qx:.3f}, {world_qy:.3f}, {world_qz:.3f}, {world_qw:.3f}]')
            
            self.get_logger().info(f'\nOdometry in Map frame:')
            self.get_logger().info(f'  Position: [{odom_map_homogeneous[0]:.3f}, '
                                  f'{odom_map_homogeneous[1]:.3f}, '
                                  f'{odom_map_homogeneous[2]:.3f}]')
            self.get_logger().info(f'  Orientation: [{odom_q_map[0]:.3f}, '
                                  f'{odom_q_map[1]:.3f}, '
                                  f'{odom_q_map[2]:.3f}, '
                                  f'{odom_q_map[3]:.3f}]')
            self.get_logger().info('=========================\n')
            
        except tf2_ros.LookupException as e:
            self.get_logger().warn(f'TF lookup failed: {e}')
        except tf2_ros.ConnectivityException as e:
            self.get_logger().warn(f'TF connectivity error: {e}')
        except tf2_ros.ExtrapolationException as e:
            self.get_logger().warn(f'TF extrapolation error: {e}')

def main(args=None):
    rclpy.init(args=args)
    node = TfConversionTest()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: test_tf_conversion.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <tf2/exceptions.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <Eigen/Geometry>

class TfConversionTest : public rclcpp::Node {
public:
  TfConversionTest() : Node("tf_conversion_test") {
    odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
        "/rkbot/lio/odom", 10, 
        std::bind(&TfConversionTest::odomCallback, this, std::placeholders::_1));

    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    RCLCPP_INFO(this->get_logger(), "TF conversion test node ready");
  }

private:
  void odomCallback(const nav_msgs::msg::Odometry::SharedPtr odom_msg) {
    // 获取 world 到 map 的变换
    Eigen::Matrix4f tf_world_to_map = Eigen::Matrix4f::Identity();
    if (!getTransform("rkbot/map", "rkbot/world", tf_world_to_map)) {
      RCLCPP_WARN(this->get_logger(), "Cannot get world->map transform, skipping");
      return;
    }

    // 从 odom 消息中提取位姿（在 world 坐标系下）
    Eigen::Matrix4f tf_world_odom = Eigen::Matrix4f::Identity();
    tf_world_odom(0, 3) = odom_msg->pose.pose.position.x;
    tf_world_odom(1, 3) = odom_msg->pose.pose.position.y;
    tf_world_odom(2, 3) = odom_msg->pose.pose.position.z;
    
    Eigen::Quaternionf q(
        odom_msg->pose.pose.orientation.w,
        odom_msg->pose.pose.orientation.x,
        odom_msg->pose.pose.orientation.y,
        odom_msg->pose.pose.orientation.z);
    tf_world_odom.block<3, 3>(0, 0) = q.toRotationMatrix();

    // 将 world 坐标系下的 odom 位姿转换到 map 坐标系
    // T_map_odom = T_world_to_map × T_world_odom
    Eigen::Matrix4f tf_map_odom = tf_world_to_map * tf_world_odom;

    // 输出结果
    RCLCPP_INFO_STREAM(this->get_logger(), "\n=== TF Conversion Test ===");
    RCLCPP_INFO_STREAM(this->get_logger(), "World->Map Transform:");
    RCLCPP_INFO_STREAM(this->get_logger(), "  Translation: [" << tf_world_to_map(0, 3) 
                    << ", " << tf_world_to_map(1, 3) << ", " << tf_world_to_map(2, 3) << "]");

    RCLCPP_INFO_STREAM(this->get_logger(), "\nOdometry in World frame:");
    RCLCPP_INFO_STREAM(this->get_logger(), "  Position: [" << odom_msg->pose.pose.position.x
                    << ", " << odom_msg->pose.pose.position.y 
                    << ", " << odom_msg->pose.pose.position.z << "]");
    RCLCPP_INFO_STREAM(this->get_logger(), "  Orientation: [" << odom_msg->pose.pose.orientation.x
                    << ", " << odom_msg->pose.pose.orientation.y 
                    << ", " << odom_msg->pose.pose.orientation.z << ", " << odom_msg->pose.pose.orientation.w << "]");

    RCLCPP_INFO_STREAM(this->get_logger(), "\nOdometry in Map frame:");
    RCLCPP_INFO_STREAM(this->get_logger(), "  Position: [" << tf_map_odom(0, 3)
                    << ", " << tf_map_odom(1, 3) << ", " << tf_map_odom(2, 3) << "]");
    
    Eigen::Quaternionf q_map(tf_map_odom.block<3, 3>(0, 0));
    RCLCPP_INFO_STREAM(this->get_logger(), "  Orientation: [" << q_map.x()
                    << ", " << q_map.y() << ", " << q_map.z() << ", " << q_map.w() << "]");
    RCLCPP_INFO_STREAM(this->get_logger(), "=========================\n");
  }

  bool getTransform(const std::string& target_frame, const std::string& source_frame, 
                    Eigen::Matrix4f& transform) {
    geometry_msgs::msg::TransformStamped transform_stamped;
    try {
      transform_stamped = tf_buffer_->lookupTransform(
          target_frame, source_frame,
          tf2::TimePointZero);
    } catch (tf2::TransformException& ex) {
      RCLCPP_WARN_STREAM(this->get_logger(), "TF lookup failed: " << ex.what());
      return false;
    }

    transform = Eigen::Matrix4f::Identity();
    transform(0, 3) = transform_stamped.transform.translation.x;
    transform(1, 3) = transform_stamped.transform.translation.y;
    transform(2, 3) = transform_stamped.transform.translation.z;

    Eigen::Quaternionf q(
        transform_stamped.transform.rotation.w,
        transform_stamped.transform.rotation.x,
        transform_stamped.transform.rotation.y,
        transform_stamped.transform.rotation.z);
    transform.block<3, 3>(0, 0) = q.toRotationMatrix();

    return true;
  }

  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TfConversionTest>());
  rclcpp::shutdown();
  return 0;
}
